MapboxVisionTests/Mocks/MockVideoSource.swift

```diff
@@ -1,10 +1,101 @@
+        let presentationTimeStamp = CMTime(value: framesAlreadyWritten, timescale: Constants.fps)
+        var sampleTimingInfo = CMSampleTimingInfo(duration: duration, presentationTimeStamp: presentationTimeStamp, decodeTimeStamp: CMTime.invalid)
+        CMSampleBufferCreateReadyWithImageBuffer(allocator: kCFAllocatorDefault,
+                                                 imageBuffer: pixelBuffer,
+                                                 formatDescription: makeFormatDescription(from: pixelBuffer)!,
+                                                 sampleTiming: &sampleTimingInfo,
+                                                 sampleBufferOut: &buffer)
+
+        framesAlreadyWritten += 1
+        let sample = VideoSample(buffer: buffer!, format: Image.Format.BGRA)
+
+        self.notify { observer in
+            observer.videoSource(self, didOutput: sample)
+        }
+    }
+
+    private func makePixelBuffer() -> CVPixelBuffer? {
+        let attrs = [
+            kCVPixelBufferMetalCompatibilityKey: kCFBooleanTrue
+        ] as CFDictionary
+        var pixelBufferUnwrapped: CVPixelBuffer?
+        let status = CVPixelBufferCreate(kCFAllocatorDefault, Int(Constants.frameWidth), Int(Constants.frameWidth), kCVPixelFormatType_32BGRA, attrs, &pixelBufferUnwrapped)
+        guard status == kCVReturnSuccess, let pixelBuffer = pixelBufferUnwrapped else {
+            return nil
+        }
+
+        CVPixelBufferLockBaseAddress(pixelBuffer, CVPixelBufferLockFlags(rawValue: 0))
+        let pixelData = CVPixelBufferGetBaseAddress(pixelBuffer)
+
+        let rgbColorSpace = CGColorSpaceCreateDeviceRGB()
+        let context = CGContext(
+            data: pixelData,
+            width: Constants.frameWidth,
+            height: Constants.frameHeight,
+            bitsPerComponent: Constants.bitsInByte,
+            bytesPerRow: CVPixelBufferGetBytesPerRow(pixelBuffer),
+            space: rgbColorSpace,
+            bitmapInfo: CGImageAlphaInfo.noneSkipFirst.rawValue
+        )
+
+        UIGraphicsPushContext(context!)
+        let value = CGFloat(sin(Float(framesAlreadyWritten) / (Float(Constants.fps) / Float.pi)) / 2.0 + 0.5)
+        let color = UIColor(red: value, green: value, blue: value, alpha: value)
+        context!.setFillColor(color.cgColor)
+        context!.fill(CGRect(x: 0, y: 0, width: Constants.frameWidth, height: Constants.frameHeight))
+        UIGraphicsPopContext()
+        CVPixelBufferUnlockBaseAddress(pixelBuffer, CVPixelBufferLockFlags(rawValue: 0))
 
-    func add(observer: VideoSourceObserver) {}
+        return pixelBuffer
+    }
 
-    func remove(observer: VideoSourceObserver) {}
+    private func makeFormatDescription(from pixelBuffer: CVPixelBuffer) -> CMFormatDescription? {
+        var formatDescription: CMFormatDescription?
+        CMVideoFormatDescriptionCreateForImageBuffer(allocator: kCFAllocatorDefault, imageBuffer: pixelBuffer, formatDescriptionOut: &formatDescription)
```

Comment:
Should we check return value?

Reply:
return value nullable, but the caller doesn't check for nil, will fix

Reply:
fixed